pkg/zcapld/verify.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package zcapld
+
+import (
+	"errors"
+	"fmt"
+)
+
+// Verifier verifies zcaps.
+type Verifier struct {
+	zcaps CapabilityResolver
+}
+
+// Proof describes the capability, the action, and the verification method of an invocation.
+type Proof struct {
+	Capability         *Capability
+	CapabilityAction   string
+	VerificationMethod string
+}
+
+// NewVerifier returns a new Verifier.
+func NewVerifier(zcapResolver CapabilityResolver) (*Verifier, error) {
+	return &Verifier{zcaps: zcapResolver}, nil
+}
+
+// Verify the proof against the invocation.
+func (v *Verifier) Verify(proof *Proof, invocation *CapabilityInvocation) error {
+	if proof.Capability == nil {
+		return errors.New(`"capability" was not found in the capability invocation proof`)
+	}
+
+	// 1. get the capability in the security v2 context
+	// **We have already resolved and parsed the full capability**
+
+	// 2. verify the capability delegation chain
+	err := v.verifyCapabilityChain(proof.Capability, proof.CapabilityAction, invocation)
+	if err != nil {
+		return fmt.Errorf("invalid capability chain: %w", err)
+	}
+
+	// 3. verify the invoker...
+	// authorized invoker must match the verification method itself OR
+	// the controller of the verification method
+	isInvoker, err := isInvoker(proof.Capability, invocation.VerificationMethod)
+	if err != nil {
+		return fmt.Errorf("isInvoke: %w", err)
+	}
+
+	if !isInvoker {
+		return errors.New("the authorized invoker does not match the verification method or its controller")
+	}
+
+	// Begin ControllerProofPurpose
+
+	// TODO the code here validates the proof's "created" time against an expected time:
+	//  nolint:lll // don't want to break the link in two
+	//  https://github.com/digitalbazaar/jsonld-signatures/blob/8d91bcb351702dde4863fab660d7ca1e5e90b2a2/lib/purposes/ProofPurpose.js#L49-L57.
+	//  Note: the higher layers in bedrock-edv-storage don't actually set `proof.created`, so `created` is always NaN.
+	//  Do we really need to verify the proof's date at this layer though? Isn't that the responsibility of a higher
+	//  layer, ie the one that parses and verifies the http signature?
+
+	// TODO verify authorization of verificationMethod.ID by controller for proof purpose `capabilityInvocation`.
+	//  Controller are probably DIDs. They have a "capabilityInvocation" property (just like DIDs) that has
+	//  verificationMethod IDs.
+
+	return nil
+}
+
+// nolint:funlen,gocyclo // TODO decompose verifyCapabilityChain into smaller units
```

Comment:
Having a GitHub issue to track these ToDos seems like a good idea to me, unless you plan on doing them soon anyway.